contracts/contracts/zapper/WOETHCCIPZapper.sol

```diff
@@ -0,0 +1,142 @@
+import { Client } from "@chainlink/contracts-ccip/src/v0.8/ccip/libraries/Client.sol";
+import { IERC20 } from "@chainlink/contracts-ccip/src/v0.8/vendor/openzeppelin-solidity/v4.8.0/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@chainlink/contracts-ccip/src/v0.8/vendor/openzeppelin-solidity/v4.8.0/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IERC4626 } from "./../../lib/openzeppelin/interfaces/IERC4626.sol";
+import { IOETHZapper } from "./../interfaces/IOETHZapper.sol";
+
+/**
+ * @title WOETH CCIP Zapper Contract
+ * @notice Hanldes logic to easily zap ETH on mainnet to WOETH on L2s like Arbitrum.
+ * @author Origin Protocol Inc
+ */
+
+contract WOETHCCIPZapper {
+    using SafeERC20 for IERC20;
+
+    /**
+     * @dev Event emitted when a zap occurs
+     * @param messageId Unique message identifier for each zap
+     * @param sender Address initiating the zap
+     * @param recipient Recipient address at destination chain
+     * @param amount Amount of ETH zapped
+     */
+    event Zap(
+        bytes32 indexed messageId,
+        address sender,
+        address recipient,
+        uint256 amount
+    );
+
+    /**
+     * @dev The destination chain selector
+     */
+    uint64 public immutable destinationChainSelector;
+
+    /**
+     * @dev The WOETH source chain (Mainnet)
+     */
+    IERC4626 public immutable woethSourceChain;
+
+    /**
+     * @dev The WOETH destination chain (Arbitrum)
+     */
+    IERC4626 public immutable woethDestinationChain;
+
+    /**
+     * @dev The OETH zapper contract address
+     */
+    IOETHZapper public immutable oethZapper;
+
+    /**
+     * @dev The CCIP router contract address
+     */
+    IRouterClient public immutable ccipRouter;
+
+    /**
+     * @dev The OETH token contract address
+     */
+    IERC20 public immutable oeth;
+
+    constructor(
+        address _ccipRouter,
+        uint64 _destinationChainSelector,
+        IERC4626 _woethSourceChain,
+        IERC4626 _woethDestinationChain,
+        IOETHZapper _oethZapper,
+        IERC20 _oeth
+    ) {
+        ccipRouter = IRouterClient(_ccipRouter);
+        destinationChainSelector = _destinationChainSelector;
+        woethSourceChain = _woethSourceChain;
+        woethDestinationChain = _woethDestinationChain;
+        oethZapper = _oethZapper;
+        oeth = _oeth;
+
+        // Max allowance for Router and WOETH contracts
+        _oeth.approve(address(_woethSourceChain), type(uint256).max); // for wrapping
+        _woethSourceChain.approve(address(_ccipRouter), type(uint256).max); // for zapping
+    }
+
+    /**
+     * @notice Accepts ETH, zaps for OETH, wraps it for WOETH and sends it to the destination chain (arbitrum)
+     * @param receiver The address of the EOA on the destination chain
+     * @param amount ETH amount to zap
+     * @return messageId The ID of the message that was sent
+     */
+    function zap(address receiver, uint256 amount)
```

Comment:
Could probably just move this to the `receive()` method. That way someone can just transfer ETH to the contract instead of invoking a method

Reply:
We could have receive forward to this method, but we certainly want to have the ability to send to a different address on the other chain, with does require having a method that can take a recipient argument.